### DISCUSSION TITLE
Text and Image, Enable Disable property!

### DISCUSSION THREAD

**Discussion by Aadil-5122** [pymupdf/PyMuPDF#2204] [Q&A]
**Is your feature request related to a problem? Please describe.**
Actually, I am working on a particular PDF in which i have to remove images/text as well as bringing them back when needed.

(Also the PDF template is same, i.e. bbox, image co-ordinates etc.)

i have successfully achieved this for the image,
but for the text, the removal what i did in first place was by "redact_annot", with that i am facing the issue that i can not bring that text back.

So, could you suggest any appropriate way to remove and bring back the text/image if any more efficient approach?

**Describe the solution you'd like**

**Describe alternatives you've considered**
Are there several options for how your request could be met?

**Additional context**

**Comment by JorjMcKie** [comment]
What did you do to get this working for images?

**Comment by Aadil-5122** [comment]
Thanks for the speedy reply,

For images, the removal was just the same as the remover.py in your repository.

For getting them back, i simply inserted it at the very "xref" from where i removed it.
using "insert_image()" function which takes in xref

as i told you that i am working on a particular type of document in which xrefs become a very handy feature (being same)

**Comment by JorjMcKie** [comment]
Ah, I understand: so you did not save the PDF after image removal (or at least not using a garbage collection).

Anyway - this is a typical "Discussions" post, not an enhancement request, so let me transfer this to the other tab.

**Reply by JorjMcKie** [reply]
There are few options to achieve what you want:

You do not seem to have a problem if things become invisible, but are still physically present. So you simply could put rectangle annotation over the places you want to temporarily hide.
To unhide, remove the annotations again.
Use rectangle annotations with white fill color and without border.

If you control the **_creation_** of your PDF, then using the "Optional Content" feature of PDF is another way. Create the images and text in question as being under control of an OCG (optional content group). In any supporting PDF viewer, this will be shown as "layers". Using PDF viewer dialog options you can dynamically (i.e. without changing the underlying PDF) toggle the visibility of objects of a layer. As an example take a look at this PDF using Adobe Acrobat or another supporting PDF reader:
[ocg-test.pdf](https://github.com/pymupdf/PyMuPDF/files/10555578/ocg-test.pdf)
There is a circle and a rectangle, and you can switch both of them on or off,
![grafik](https://user-images.githubusercontent.com/8290722/216007965-20352072-8091-4187-9d89-11a17c97b4cb.png)

**Comment by Aadil-5122** [comment]
during the process, i am completely ok if the "elements" are invisible, is there any wrapper function which i can create, which removes the invisible elements or so? or any particular saving technique?

**Reply by Aadil-5122** [reply]
Basically like once removed or added, a particular element, then a wrapper function which considers foreground of the dhaar card as the new state and hence saves it that way?

**Reply by JorjMcKie** [reply]
Are you referring to using Optional Content?
Adding images, annotation, text or drawings can be done in a way that uses Optional Content parameters (i.e. OCGs, optional content groups).
This makes it possible to switch the visibility on or off for each element under the control of an OCG. The visibility state can be done temporarily within the running process, or also permanently stored when saving the PDF - in which case everything is visible / invisible like when it was saved.

**Reply by Aadil-5122** [reply]
Alright Sir, Thanks alot!

**Reply by Aadil-5122** [reply]
Also sir, i am not sure whether this question fits here in this discussion or not:

but, i would like to ask that, after i have enabled or disabled an image, 

can i use the "updated" document further without saving it!

For example:
i start with "doc1"

after i have disabled a particular image in it, 
now can this updated document (without the image) be used further without first saving it to the locals and then re-accessing it?

**Reply by Aadil-5122** [reply]
simply in the form of a variable!
